Issue #68 Refactoring for JDK6 support

tolbertam opened the pull request:
Removes cassandra-all dependency and replaces code that used cassandra-all to use custom equivalents so Scassandra can be used with JDK6.

Committing now just to get something up to look at.  There are currently some issues with collection serialization.  I suspect I have some issues around how i'm sizing elements in collections.  There was code that wrote types differently depending on whether or not in a collection (i.e. text wrote a 4-byte length header before each string when not in a collection, 2 byte otherwise) that I removed, but it looks like that may be necessary (not sure why, can't see anything in spec differentiating to two).

chbatey commented:
Will start looking at this tomorrow. Can you start including the issue number in commits, I try to do this but most of the time I forget

tolbertam commented on java-client/src/main/java/org/scassandra/http/client/PrimingRequest.java:
```diff
@@ -66,7 +66,6 @@ public PrimingRequestBuilder withRows(List<Map<String, ?>> rows) {
             return this;
         }
 
-        @SafeVarargs
```
SafeVarargs and Objects.hash both require Java7.  Replaced objects.hash with autogenerated equals/hashCode methods in IntelliJ

tolbertam commented on java-it-tests/common/src/main/java/queries/PrimingListsForQuery.java:
```diff
@@ -57,7 +57,7 @@ public void testVarcharTextAndAsciiLists() {
                 "list_type_ascii", ColumnTypes.AsciiList
         );
-        ImmutableMap<String, CqlType> columnTypesAsNewCqlType = ImmutableMap.of(
+        ImmutableMap<String, ListType> columnTypesAsNewCqlType = ImmutableMap.of(
```
Needed to be more explicit with types with JDK6 for some reason.

chbatey replied:
yeah type inference for generics got better in 7

tolbertam commented on server/src/main/scala/org/scassandra/server/cqlmessages/CqlProtocolHelper.scala:
```diff
@@ -18,17 +18,37 @@ package org.scassandra.server.cqlmessages
+    val bytes = f(value)
+    protocolVersion match {
+      // proto <= v2 uses short length for elements, >= 3 uses int length.
+      case VersionOne | VersionTwo => withShortLength(bytes)
```
I do this in a number of places and it should probably be generalized (will work on that).

tolbertam commented on server/src/main/scala/org/scassandra/server/cqlmessages/types/ColumnType.scala:
```diff
@@ -15,21 +15,26 @@
+  implicit val byteOrder = java.nio.ByteOrder.BIG_ENDIAN
+  def readValue(byteIterator : ByteIterator)(implicit protocolVersion: ProtocolVersion) : Option[T]
+  def writeValue(value : Any)(implicit protocolVersion: ProtocolVersion) : Array[Byte]
+  def readValueWithLength(byteIterator: ByteIterator)(implicit protocolVersion: ProtocolVersion): Option[T] = {
```
Since depending on protocolVersion and context (whether or not value will be part of a collection) matters when encoding and decoding length I've decided to divorce length from read and writeValue.  Instead the byteIterator passed into readValue will only encompass the bytes for that data (so code can make an assumption about length).  Additionally, code that calls writeValue must encode the length on the outside now (CqlProtocolHelper.withLength provided for this, and collection serializers do this too).  readValueWithLength is a convenience method which takes a ByteIterator and gets the necessary slice of data based on read int length and passes the the slice to readValue (I'll document this).

tolbertam replied:
Additionally, there's some inefficiencies I think i've introduced in code going back and forth between ByteIterator and Array[Byte].  I'll look to clean that up a bit.

chbatey replied:
I think we have a lot of those and I am to blame for most of them :)

chbatey commented on server/build.gradle:
```diff
@@ -120,7 +120,6 @@ dependencies {
     compile  "io.spray:spray-routing_2.11:1.3.3"
     compile  "com.typesafe.scala-logging:scala-logging_2.11:3.1.0"
     compile  "com.google.guava:guava:17.0"
-    compile  "org.apache.cassandra:cassandra-all:2.1.2" //).exclude("org.antlr", "stringtemplate")
```
+1 Good bye!

chbatey commented on server/src/main/scala/org/scassandra/server/cqlmessages/CqlProtocolHelper.scala:
```diff
@@ -59,99 +79,87 @@ object CqlProtocolHelper {
+      case VersionOne | VersionTwo => serializeCollection(_.putShort(_))
+      case VersionThree => serializeCollection(_.putInt(_))
+    }
+    serializer(input, entrySerializer)
```
Maybe we could have this as a class rather than a object and have different implementations for the various versions. Would then be very explicit when they differ. Like with the CqlMessagefactory

tolbertam replied:
+1 on that idea, also that should eliminate the need to pass an implicit protocolVersion.  I have a change coming that moves some of the protocol checking around, but still more to do there.

tolbertam replied:
Made a change that gets rid of the protocol version checking for determining collection length.   Added ProtocolVersion.collectionLength.  Not completely satisfied with the solution but solves my immediate problem.

tolbertam commented:
I think this is in a somewhat complete state now so I removed [work in progress].   After review and if everything looks ok I'll squash commits and update the commit message to include issue #.

chbatey commented on build.gradle:
```diff
@@ -4,6 +4,12 @@ plugins {
+    apply plugin: 'java'
+    sourceCompatibility = 1.6
+    targetCompatibility = 1.6
+}
```
Do you think it is worthwhile just applying this to server/java-client so we can use java 8 in the it-tests?

tolbertam replied:
That sounds good to me, i'll make that update.

chbatey commented on server/src/main/scala/org/scassandra/server/cqlmessages/ProtocolVersion.scala:
```diff
@@ -35,7 +37,27 @@ object ProtocolVersion {
+ * on a byte stream.
+ */
+trait CollectionLength {
+  implicit val byteOrder = java.nio.ByteOrder.BIG_ENDIAN
```
import from CqlProtocolHelper so we only have it once

tolbertam replied:
Good call, I think I added this a couple other places as well. Will fix.

chbatey commented on server/src/main/scala/org/scassandra/server/cqlmessages/request/ExecuteRequest.scala:
```diff
@@ -46,7 +46,7 @@ object ExecuteRequest {
 
     val variableValues = variableTypes.map (varType => {
-      varType.readValue(bodyIterator, VersionTwo)
+      varType.readValueWithLength(bodyIterator)(VersionTwo)
```
Reminding me of the nastyness I did with methods for each version rather than use the CqlMessageFactory :(

tolbertam replied:
Yeah, I still think there is room for improvement here.    I switched protocol version to implicit so it didn't have to be passed around when calling readValue from a Collection ColumnType.   I've now isolated all explicit protocol checks in ColumnType/CqlProtocolHelper to making calls to ProtocolVersion (added CollectionLength for that).

chbatey commented on server/src/test/scala/org/scassandra/server/cqlmessages/types/CqlTimestampTest.scala:
```diff
@@ -17,13 +17,7 @@ package org.scassandra.server.cqlmessages.types
-    deserialisedValue should equal(None)
-  }
+class CqlTimestampTest extends FunSuite with Matchers with ProtocolProvider {
 }
```
I guess either add some non-null cases here or remove the test

tolbertam replied:
Good catch, I'll add a test case.

chbatey commented on .travis.yml:
```diff
@@ -2,7 +2,8 @@ language: scala
 jdk:
   - oraclejdk8
```
Should we just run this in travis with 6 to make sure?

tolbertam replied:
Based on our other comment about only java-client and server targeting 1.6 source compatibility, should we still consider running with jdk6?

chbatey replied:
Just concerned we may not catch a java 7 jar sneaking in, tho I guess your Java Driver build will break :)

tolbertam replied:
Ahhh, that is a good point!  There is definitely a difference between compile and runtime.  Maybe there is some travis magic we can do to run the server and client tests at JDK 6 and the java-it-tests at JDK8.  Does that sound like a good route?  I can look into that.

tolbertam replied:
I should also add some more tests for collections.   With the changes here, we should in theory be able to create collections of any cql type (including nesting collections (I think)).  I'll do that.

chbatey replied:
Looks like you can:

script:
  - jdk_switcher use oraclejdk8
  - # do stuff with Java 8
  - jdk_switcher use oraclejdk7
  - # do stuff with Java 7

http://docs.travis-ci.com/user/languages/java/

tolbertam replied:
Nice!  That's pretty neat, i'll give it a go.

tolbertam replied:
I reverted the commit that changed java-it-tests so it supported jdk6.   Also updated build.gradle to remove source/target compatibility and project root level.  Travis will now build with jdk8, run the java-it-tests at jdk8 and the rest of the tests at jdk6.

tolbertam commented:
I should also add some more tests for collections.   With the changes here, we should in theory be able to create collections of any cql type (including nesting collections (I think)).  I'll do that.

chbatey commented:
@tolbertam Much more to do on this? Be great to get it merged.

tolbertam commented:
@chbatey Sorry for the delay here, Its very close to being done, I just need to add a few more tests and then clean up the commit history a bit.  I'll finish up the tests this morning and then if everything looks good to you I'll clean up the commit history.

tolbertam commented:
Just noticed that there are some changes i made in java-it-tests that are no longer required since we opted to keep that requiring JDK 8.   I'm working on backing out those particular changes and will do another push.

tolbertam commented:
Apologies for the storm of notifications this must have generated.  Everything is looking good, just need to do some rebasing on master to resolve some conflicts, should have that fixed shortly.

tolbertam commented:
The only test that is failing is the ActivityClient#metadataPriming test because it hits a socket timeout (500ms).  It runs fine in my environment.  I noticed that the ActivityClient sets explicit timeouts, but the PrimingClient does not.  Is there a reason for that?  I tried removing it, but it looks like there are tests depending on those timeouts, so I left it as is.

I think everything should be good to go now.

tolbertam commented:
Added another commit because I noticed the following exception happening in the logs:

    [ERROR] [07/09/2015 17:58:26.816] [CassandraServerStub-8042-8043-akka.actor.default-dispatcher-9] [akka://CassandraServerStub-8042-8043/user/BinaryTcpListener/$e/$b] next on empty iterator
    java.util.NoSuchElementException: next on empty iterator
    	at scala.collection.Iterator$$anon$2.next(Iterator.scala:39)
    	at scala.collection.Iterator$$anon$2.next(Iterator.scala:37)
    	at akka.util.ByteIterator$ByteArrayIterator.next(ByteIterator.scala:40)
    	at akka.util.ByteIterator.getShort(ByteIterator.scala:477)
    	at org.scassandra.server.cqlmessages.AbstractMessageFactory.parseQueryRequest(AbstractMessageFactory.scala:89)
    	at org.scassandra.server.actors.QueryHandler$$anonfun$receive$1.applyOrElse(QueryHandler.scala:63)
    	at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
    	at org.scassandra.server.actors.QueryHandler.aroundReceive(QueryHandler.scala:26)
    	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
    	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
    	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
    	at akka.dispatch.Mailbox.run(Mailbox.scala:221)
    	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
    	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

This didn't cause any test failures or problems as is.  I updated the code to handle the case where no values are present in a QueryRequest by checking the flags.

chbatey commented:
Regarding the MetaDataPriming20

I think both the ActivityClient and the PrimingCline should set the timeouts and they probably should be configurable. I'll raise a ticket for that, feel free to increase the timeout and adjust the test if need be.

I take it you still need to reabse against master?

chbatey commented on .travis.yml:
```diff
@@ -2,7 +2,12 @@ language: scala
+  - jdk_switcher use openjdk6
+  - ./gradlew check -x signArchives -x java-it-tests:common:check -x java-it-tests:driver20:check -x java-it-tests:driver21:check
+  - jdk_switcher use oraclejdk8
+  - ./gradlew java-it-tests:common:check java-it-tests:driver20:check java-it-tests:driver21:check -x signArchives
```
travis is pretty cool!

tolbertam replied:
Yeah, I am a fan!  Seems like it has an answer for everything :).  The only thing I couldn't wrap my head around was finding a way to omit a module and it's children completely when building with gradle.  For now i'm excluding the check phase for each module individually in jdk6 which seems to work (`-x java-it-tests:common:check -x java-it-tests:driver20:check -x java-it-tests:driver21:check`)

chbatey commented on server/src/main/scala/org/scassandra/server/cqlmessages/AbstractMessageFactory.scala:
```diff
@@ -86,10 +86,15 @@ abstract class AbstractMessageFactory extends CqlMessageFactory {
+        })
+      }
+      case _   => Nil
+    }
```
I added QueryFlagParser when doing query parameters - is this the same flag? Hard to tell in GitHub

tolbertam replied:
Ah nice, yep it's the same thing I believe.  It doesn't look like QueryFlagParser is used except for Testing, but it would apply perfectly here.  I'll change it to use it.

tolbertam replied:
Updated to use QueryFlagParser

chbatey commented on server/src/main/scala/org/scassandra/server/cqlmessages/CqlProtocolHelper.scala:
```diff
@@ -43,116 +48,85 @@ object CqlProtocolHelper {
+   * @tparam U type of elements being encoded.
+   * @return The encoded bytes.
+   */
+  def serializeCollection[U](sizeF: (ByteStringBuilder, Int) => ByteStringBuilder)(input: TraversableOnce[U], entrySerializer: U => Array[Byte]) : Array[Byte] = {
```
Do you think we should attach the sizeF to either the ProtocolVersion class or something? No problem is this a pain and don't have it here. Isolating the differences will become more important the more we add

tolbertam replied:
It depends on how this method is used.  This is called by CqlProtoclHelper.serializeStringCollection (which is used for [string list], [string map] and [string multimap]) in which it provides putShort (same for all protocol versions), and it's also used by ColumnType.serializeCqlCollection which uses protocolVersion.collectionLength.getLength.

chbatey commented on server/src/main/scala/org/scassandra/server/cqlmessages/response/Response.scala:
```diff
@@ -28,7 +30,7 @@ abstract class Response(header : Header) extends CqlMessage(header)
 case class Ready(stream : Byte = ResponseHeader.DefaultStreamId)(implicit protocolVersion: ProtocolVersion) extends Response(new Header(protocolVersion.serverCode, opCode = OpCodes.Ready, streamId = stream)) {
 
-  implicit val byteOrder = java.nio.ByteOrder.BIG_ENDIAN
+  implicit val byteOrder : ByteOrder = CqlProtocolHelper.byteOrder
```
You could just import CqlProtocolHelper._ , no preference just which ever you think is nicer.

tolbertam replied:
It looks like I missed this one, I had fixed in other places but must have missed here.

tolbertam replied:
Oh, now I see what you mean, yeah I'll do that instead,

tolbertam replied:
Fixed in all applicable places.

chbatey commented on server/src/main/scala/org/scassandra/server/cqlmessages/types/CqlTextType.scala:
```diff
@@ -0,0 +1,16 @@
+import akka.util.ByteIterator
+import org.scassandra.server.cqlmessages.ProtocolVersion
+
+abstract class CqlTextType(override val code : Short, override val stringRep: String, val charSet: String = "UTF-8") extends ColumnType[String](code: Short, stringRep: String) {
```
:+1: Nice

tolbertam commented:
I think I need to rebase against master again since travis PR was merged, should be simple.  I'll do that quick and resolve some of your other comments since I missed a couple of things.  Should be done quickly!

tolbertam commented:
I think i've addressed everything but if I missed anything let me know :smile:

chbatey commented:
Want to squash the commits?

tolbertam commented:
ah yes, I should probably take care of those two extra commits (d34efff and 51d4152), i'll squash them into 176c337.  Would you prefer that I squash all the commits to 1, or keep 4 separate commits (was originally trying to break them up by module effected, but maybe that isn't necessary)?

chbatey commented:
Which ever you prefer :)

tolbertam commented:
Awesome, thanks.  I went with the 4 commits, just pushed something up, hope that looks ok!

chbatey commented:
Any ideas on the failures?

tolbertam commented:
It looks like just21.MetaDataPriming21 is caused by the SocketTimeout configuration on ActivityLog (discussed further up in this PR).  The other two i'm not exactly sure of yet.  Looking into it.

tolbertam commented:
Can't seem to get them to fail locally, but will keep looking.  Just for my sanity, could you trigger travis to rerun the job?

chbatey commented:
I tried as well. We need the timeout to be configurable and increased it a tad for travis. Can raise a separate issue for that.

tolbertam commented:
Turns out that I can reproduce this locally running 'gradle :java-it-tests:driver21:test' from the project root, debugging further.

tolbertam commented:
Oh i see,  this is ringing some bells, there was a recent change for 2.1.7/2.0.11 that changes the default behavior of RetryPolicy for UnavailableException to try on the next host at most once.

> 'com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: localhost/127.0.0.1:8042 (com.datastax.driver.core.exceptions.UnavailableException: Not enough replica available for query at consistency LOCAL_ONE (4 required but only 3 alive)))'

tolbertam commented:
From the [upgrade guide for 2.1.7](http://datastax.github.io/java-driver/2.1.7/upgrade_guide/#2-1-7):

> The DefaultRetryPolicy’s behaviour has changed in the case of an Unavailable exception received from a request. The new behaviour will cause the driver to process a Retry on a different node at most once, otherwise an exception will be thrown. This change makes sense in the case where the node tried initially for the request happens to be isolated from the rest of the cluster (e.g. because of a network partition) but can still answer to the client normally. In this case, trying another node has a chance of success. The previous behaviour was to always throw an exception.

Since 2.1.7 just released today, the test started failing.  We have a similar test in the java-driver (which uses scassandra) that we added for this and had to change some others (see [this](https://github.com/datastax/java-driver/pull/391)).  I'll update the tests to reflect the behavior change.

chbatey commented:
That's amusing!



> On 15 Jul 2015, at 20:55, Andrew Tolbert <notifications@github.com> wrote:
> 
> From the upgrade guide for 2.1.7:
> 
> The DefaultRetryPolicy’s behaviour has changed in the case of an Unavailable exception received from a request. The new behaviour will cause the driver to process a Retry on a different node at most once, otherwise an exception will be thrown. This change makes sense in the case where the node tried initially for the request happens to be isolated from the rest of the cluster (e.g. because of a network partition) but can still answer to the client normally. In this case, trying another node has a chance of success. The previous behaviour was to always throw an exception.
> 
> Since 2.1.7 just released today, the test started failing. We have a similar test in the java-driver (which uses scassandra) that we added for this and had to change some others (see this). I'll update the tests to reflect the behavior change.
> 
> —
> Reply to this email directly or view it on GitHub.
>

tolbertam commented:
I think the easy fix for this is to just change the RetryPolicy to not retry on another host on unavailable.  I'll make that change and push it up if it works good in both the 2.0 and 2.1 cases.

tolbertam commented:
Closed/reopened to retrigger a build (not sure why the previous test failed, can't reproduce locally).   I think this should be good to go now, but let me know and i'd be happy to fix :).
